backend/server/routers/programs.py

```diff
@@ -78,7 +78,7 @@ def add_subgroup_container(structure: dict, type: str, container: dict, exceptio
     title = container.get("title")
     if container.get("type") == "gened":
         title = "General Education"
-    if container.get("type") is not None and "rule" in container.get("type"):
+    if container.get("type") is not None and "rule" in container.get("type"): # type: ignore
```

Comment:
Why does this need an ignore.
i.e - why does it think this type is and what do we actaully want it to be

Reply:
Yea this is ignoring the case where it is None. For some reason, mypy doesn't understand short circuit

Reply:
hmmm, sounds like a consequence of dictionary w/out strongly defined types and classes (OOP and its consequence :( )
To avoid the `#type: ignore`, we could possibly just do this with a `supress(KeyError)`

```py
type = "whatever_default"
with suppress(KeyError):
  if "rule" in container.get("type"):
    type = "Rule"
```

But just leaving this as is, is fine too tbh

Reply:
N.B: I haven't read the code

but, if you're using mypy (or any typechecker) you should almost never have #ignores - unless there is literally a bug in the type system, or no reasonable workaround.

Have you investigated having `container` being a `TypedDict` https://mypy.readthedocs.io/en/stable/more_types.html#typeddict

Reply:
Aside: walrus is fun here if you want to minimise the number of indexes you do
```py3
if (type := container.get("type")) is not None and "rule" in type :
```

Reply:
ooo this was a very nice rabbithole, which helps heaps with accessing the database. I typedicted the database results now, so should be cool 👀